### ISSUE TITLE
Auto generate tweakme.h

### ISSUE THREAD

**Issue by xvitaly** [gabime/spdlog#3077]
It looks like the `include/spdlog/tweakme.h` file should be automatically generated by CMake depending on the build configuration.

Various Linux distributions need to manually patch it in downstream to resolve issues with packages not using pkg-config or CMake.

**Comment by tt4g** [comment]
> It looks like the `include/spdlog/tweakme.h` file should be automatically generated by CMake depending on the build configuration.

CMake shouldn't generate `tweakme.h`, what makes you think it does?

**Comment by xvitaly** [comment]
> CMake shouldn't generate tweakme.h, what makes you think it does?

I think it should be auto-generated for feature parity with pkg-config and CMake configurations.

**Comment by tt4g** [comment]
Sorry. I am not a native speaker so I misunderstood the meaning.

spdlog was initially a header-only library. The option to build as a shared or static library was added later.
I believe that when it was header-only, there was no automatic `tweakme.h` generation because there was no significant difference between downloading and installing.
It is technically possible to generate `tweakme.h` at build time.
However, this may cause inconvenience for some users who still use spdlog as a header-only library and they have multiple projects using spdlog (even though spdlog is installed as a library, users can use it as a header-only library with a CMake target named `spdlog::spdlog_header_only`).

The final decision must be made by the maintainer @gabime.

**Comment by xvitaly** [comment]
> However, this may cause inconvenience for some users who still use spdlog as a header-only library and they have multiple projects using spdlog (even though spdlog is installed as a library, users can use it as a header-only library with a CMake target named spdlog::spdlog_header_only).

Auto-generation can be enabled only if the `SPDLOG_BUILD_SHARED` flag is set.

**Comment by tt4g** [comment]
`spdlog::spdlog_header_only` target is added with or without the `SPDLOG_BUILD_SHARED` flag.
When installed by CMake, spdlog can be either header-only and static/shared library.

**Comment by xvitaly** [comment]
>  When installed by CMake, spdlog can be either header-only and static/shared library.

Yes, but it has major issues with packages not using pkg-config or CMake. For more information please check [this ticket](https://bugzilla.redhat.com/show_bug.cgi?id=2249227).

**Comment by gabime** [comment]
I’ve been working on something like this in the v2.x branch.  cmake would generate a [config.h](https://github.com/gabime/spdlog/blob/v2.x/cmake/spdlog_config.h.in) file. Not sure yet if this the right approach though..

**Comment by tt4g** [comment]
If this is considered a major change, it would make sense to automatically generate a configuration file from v2.
I personally think it is better to keep the style of macro definition by the developer as before, as it may confuse some users to automatically generate `tweakme.h` in v1.

**Comment by xvitaly** [comment]
> cmake would generate a [config.h](https://github.com/gabime/spdlog/blob/v2.x/cmake/spdlog_config.h.in) file. Not sure yet if this the right approach though..

Looks good. Is it possible to backport this to 1.x?

**Comment by gabime** [comment]
Porting it to v1.x would break current usage of tweakme.h or add confusion if generate a new file in addition to tweakme.h.

**Comment by xvitaly** [comment]
> Porting it to v1.x would break current usage of tweakme.h or add confusion if generate a new file in addition to tweakme.h.

Even if it will be optional and disabled by default? `SPDLOG_GENERATE_CONFIG` for example. This feature can be very useful for Linux package maintainers.

If enabled, the `config.h` file will be generated and installed instead of `tweakme.h`.

**Comment by gabime** [comment]
How would the new config.h be included by the code? Also note that when consuming spdlog 
using cmake find_ pacakge, cmake already adds the right defines(e.g. `target_compile_definitions(spdlog PUBLIC SPDLOG_SHARED_LIB)` ,so what would happen if both methods used. Still seems confusing.

**Comment by tt4g** [comment]
If a macro was defined when the library was built, CMake will define the same macro for the application to be linked.
If the same macro is also defined in `config.h`, the macro definitions will be duplicated, but as long as the definitions are the same, this should not be a problem.

**Comment by xvitaly** [comment]
> How would the new config.h be included by the code?

Something like that:
```c++
#if __has_include(<spdlog/config.h>)
#include <spdlog/config.h>
#else
#include <spdlog/tweakme.h>
#endif
```

> Also note that when consuming spdlog using cmake find_ pacakge, cmake already adds the right defines(e.g. target_compile_definitions(spdlog PUBLIC SPDLOG_SHARED_LIB) ,so what would happen if both methods used. Still seems confusing.

Yes, but we're talking about a [situation](https://bugzilla.redhat.com/show_bug.cgi?id=2249227) where the packages don't use pkg-config or CMake.

**Comment by gabime** [comment]
__has_include is c++17 while spdlog v1.x is c++11

**Comment by xvitaly** [comment]
> __has_include is c++17 while spdlog v1.x is c++11

Another option is to install the generated `config.h` file under the name `tweakme.h` if the `SPDLOG_GENERATE_CONFIG` flag is enabled.

**Comment by gabime** [comment]
Perhaps a `SPDLOG_GENERATE_TWEAKME_H`cmake option (default=OFF) is indeed good way to solve this for 1.x.

PR is welcme.
